episodes/5-transfer-learning.Rmd

```diff
@@ -0,0 +1,235 @@
+
+An example: Let's say that you want to train a model to classify images of different dog breeds. You could make use of a pre-trained network that learned how to classify images of dogs and cats. The pre-trained network will not know anything about different dog breeds, but it will have captured some general knowledge of, on a high-level, what dogs look like, and on a low-level all the different features (eyes, ears, paws, fur) that make up an image of a dog. Further training this model on your dog breed dataset is a much easier task than training from scratch, because the model can use the general knowledge captured in the pre-trained network.
+
+In this episode we will learn how use Keras to apply a state-of-the-art pre-trained model to the dollarstreet-10 dataset.
+
+
+## 1. Formulate / Outline the problem and 2. Identify inputs and outputs
+
+
+Just like in the previous episode, we use the dollar street 10 dataset. 
+The goal is to predict one out of 10 classes for a given image.
+
+We load the data in the same way as the previous episode:
+```python
+import pathlib
+import numpy as np
+
+DATA_FOLDER = pathlib.Path('data/dataset_dollarstreet/') # change to location where you stored the data
+train_images = np.load(DATA_FOLDER / 'train_images.npy')
+val_images = np.load(DATA_FOLDER / 'test_images.npy')
+train_labels = np.load(DATA_FOLDER / 'train_labels.npy')
+val_labels = np.load(DATA_FOLDER / 'test_labels.npy')
+```
+
+## 3. Prepare the data
+We prepare the data as before, scaling the values between 0 and 1.
+```python
+train_images = train_images / 255.0
+val_images = val_images / 255.0
+```
+
+## 4. Choose a pre-trained model or start building architecture from scratch
+Let's define our model input layer using the shape of our training images:
+```python
+# input tensor
+inputs = keras.Input(train_images.shape[1:])
+```
+
+Our images are 64 x 64 pixels, whereas the pre-trained model that we will use was 
+trained on images of 160 x 160 pixels. To deal with this, we add an upscale layer
+that resizes the images to 160 x 160 pixels during training and prediction.
+```python
+# upscale layer
+method = tf.image.ResizeMethod.BILINEAR
+upscale = keras.layers.Lambda(
+  lambda x: tf.image.resize_with_pad(x, 160, 160, method=method))(inputs)
+```
+
+From the `keras.applications` module we use the `DenseNet121` architecture. 
+This architecture was proposed by the paper: [Densely Connected Convolutional Networks (CVPR 2017)](https://arxiv.org/abs/1608.06993). It is trained on the [Imagenet](https://www.image-net.org/) dataset, which contains 14,197,122 annotated images according to the WordNet hierarchy with over 20,000 classes.
+
+We will have a look at the architecture later, for now it is enough to know
+that it is a convolutional neural network with 121 layers that was designed 
```

Comment:
I wonder: is it the large amount of data or the large number of parameter that leads to the improved accuracy? Or: is it really the transfer learning aspect that gives the boost?
My guess is that it is both, but I think that we should have a more specific answer.